ert_shared/ensemble_evaluator/dispatch.py

```diff
@@ -41,27 +48,86 @@ async def join(self):
 
 
 class Dispatcher:
-    def __init__(self, batcher=None):
-        self.__LOOKUP_MAP = defaultdict(list)
+    def __init__(self, ensemble, evaluator_callback, batcher=None):
+        self._LOOKUP_MAP = defaultdict(list)
         self._batcher: Optional[Batcher] = batcher
-
-    def register_event_handler(self, event_types, batching=False):
-        def decorator(function):
-            nonlocal event_types, batching
-            if not isinstance(event_types, set):
-                event_types = set({event_types})
-            for event_type in event_types:
-                self.__LOOKUP_MAP[event_type].append((function, batching))
-
-            def wrapper(*args, **kwargs):
-                return function(*args, **kwargs)
-
-            return wrapper
-
-        return decorator
+        self._ensemble = ensemble
+        self._evaluator_callback = evaluator_callback
+        self._register_event_handlers()
+
+    def register_event_handler(self, event_types, function, batching=False):
+        if not isinstance(event_types, set):
+            event_types = {event_types}
+        for event_type in event_types:
+            self._LOOKUP_MAP[event_type].append((function, batching))
+
+    def _register_event_handlers(self):
+        self.register_event_handler(
+            event_types=identifiers.EVGROUP_FM_ALL,
+            function=self._fm_handler,
+            batching=True,
+        )
+        self.register_event_handler(
+            event_types=identifiers.EVTYPE_ENSEMBLE_STOPPED,
+            function=self._ensemble_stopped_handler,
+            batching=True,
+        )
+        self.register_event_handler(
+            event_types=identifiers.EVTYPE_ENSEMBLE_STARTED,
+            function=self._ensemble_started_handler,
+            batching=True,
+        )
+        self.register_event_handler(
+            event_types=identifiers.EVTYPE_ENSEMBLE_CANCELLED,
+            function=self._ensemble_cancelled_handler,
+            batching=True,
+        )
+        self.register_event_handler(
+            event_types=identifiers.EVTYPE_ENSEMBLE_FAILED,
+            function=self._ensemble_failed_handler,
+            batching=True,
+        )
+
+    async def _fm_handler(self, events):
+        snapshot_update_event = self._ensemble.update_snapshot(events)
+        await self._evaluator_callback(
+            identifiers.EVTYPE_EE_SNAPSHOT_UPDATE, snapshot_update_event
+        )
+
+    async def _ensemble_stopped_handler(self, events):
+        if self._ensemble.get_status() != ENSEMBLE_STATE_FAILED:
+            await self._evaluator_callback(
+                identifiers.EVTYPE_ENSEMBLE_STOPPED,
+                self._ensemble.update_snapshot(events),
+                events[0].data,  # TODO fix this
```

Comment:
Fix what?

Reply:
Fix the fact that we are forced to get the element inside a list with only one element

Reply:
And that still needs a fix :disappointed: